pilot/pkg/networking/core/v1alpha3/cluster_builder.go

```diff
@@ -566,24 +566,16 @@ func (cb *ClusterBuilder) buildLocalityLbEndpoints(proxyView model.ProxyView, se
 			Metadata: &core.Metadata{},
 		}
 
-		labels := instance.Endpoint.Labels
-		ns := instance.Endpoint.Namespace
+		metadata := instance.Endpoint.Metadata()
 		if features.CanonicalServiceForMeshExternalServiceEntry && service.MeshExternal {
-			ns = service.Attributes.Namespace
+			metadata.Namespace = service.Attributes.Namespace
 			svcLabels := service.Attributes.Labels
 			if _, ok := svcLabels[model.IstioCanonicalServiceLabelName]; ok {
-				labels = map[string]string{
-					model.IstioCanonicalServiceLabelName:         svcLabels[model.IstioCanonicalServiceLabelName],
-					model.IstioCanonicalServiceRevisionLabelName: svcLabels[model.IstioCanonicalServiceRevisionLabelName],
-				}
-				for k, v := range instance.Endpoint.Labels {
-					labels[k] = v
-				}
+				metadata.Labels[model.IstioCanonicalServiceLabelName] = svcLabels[model.IstioCanonicalServiceLabelName]
```

Comment:
This causes a concurrent map write from one gothreads creating xDS:

```
fatal error: concurrent map writes

goroutine 1751 [running]:
istio.io/istio/pilot/pkg/networking/core/v1alpha3.(*ClusterBuilder).buildLocalityLbEndpoints(0xc001d2f800, {0x493b8c8, 0x70bc400}, 0xc002bfe2c0, 0xc0040a8480?, 0xc001f57980?)
	istio.io/istio/pilot/pkg/networking/core/v1alpha3/cluster_builder.go:578 +0x5c6
istio.io/istio/pilot/pkg/networking/core/v1alpha3.(*ConfigGeneratorImpl).buildOutboundClusters(0xc0006423c0?, 0xc001d2f800, 0x495dce0?, {0xc002f1c450?, 0x18?}, {0xc00309eb80, 0xd, 0x7fa2a444e9f8?})
	istio.io/istio/pilot/pkg/networking/core/v1alpha3/cluster.go:257 +0x3f9
istio.io/istio/pilot/pkg/networking/core/v1alpha3.(*ConfigGeneratorImpl).buildClusters(0xc0011b87c0, 0xc0006423c0, 0xc0041ada40, {0xc00309eb80, 0xd, 0x10})
	istio.io/istio/pilot/pkg/networking/core/v1alpha3/cluster.go:162 +0x13e
istio.io/istio/pilot/pkg/networking/core/v1alpha3.(*ConfigGeneratorImpl).BuildClusters(0x38e24e0?, 0xc0007259e0?, 0x46a6?)
	istio.io/istio/pilot/pkg/networking/core/v1alpha3/cluster.go:88 +0x90
istio.io/istio/pilot/pkg/xds.CdsGenerator.Generate({0x0?}, 0xc0006423c0?, 0xc000c87380?, 0xc0041ada40?)
	istio.io/istio/pilot/pkg/xds/cds.go:80 +0xb5
istio.io/istio/pilot/pkg/xds.(*DiscoveryServer).pushXds(0xc0011ac780, 0xc002f8e280, 0xc0006bff80, 0xc0041ada40)
	istio.io/istio/pilot/pkg/xds/xdsgen.go:121 +0x2d7
istio.io/istio/pilot/pkg/xds.(*DiscoveryServer).pushConnection(0xc0011ac780, 0xc002f8e280, 0xc002f8e280?)
	istio.io/istio/pilot/pkg/xds/ads.go:757 +0x1a5
istio.io/istio/pilot/pkg/xds.(*DiscoveryServer).Stream(0xc0011ac780, {0x495f490, 0xc002853780})
	istio.io/istio/pilot/pkg/xds/ads.go:340 +0xa70
istio.io/istio/pilot/pkg/xds.(*DiscoveryServer).StreamAggregatedResources(0xc002cdbb30?, {0x495f490?, 0xc002853780?})
	istio.io/istio/pilot/pkg/xds/ads.go:237 +0x25
github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3._AggregatedDiscoveryService_StreamAggregatedResources_Handler({0x407da40?, 0xc0011ac780}, {0x495a510?, 0xc002fd8690})
	github.com/envoyproxy/go-control-plane@v0.11.1-0.20230414193417-178e253923d1/envoy/service/discovery/v3/ads.pb.go:298 +0x9f
google.golang.org/grpc.(*Server).processStreamingRPC(0xc000e823c0, {0x4962ca0, 0xc001c551e0}, 0xc0014490e0, 0xc001ce6ab0, 0x6dabd20, 0x0)
	google.golang.org/grpc@v1.54.0/server.go:1639 +0x1384
google.golang.org/grpc.(*Server).handleStream(0xc000e823c0, {0x4962ca0, 0xc001c551e0}, 0xc0014490e0, 0x0)
	google.golang.org/grpc@v1.54.0/server.go:1726 +0x9f0
google.golang.org/grpc.(*Server).serveStreams.func1.2()
	google.golang.org/grpc@v1.54.0/server.go:966 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
	google.golang.org/grpc@v1.54.0/server.go:964 +0x28a
```

Reply:
Which two threads are writing concurrently